File: src/test/java/com/joecollins/graphics/components/GraphicsFrameTest.java
package com.joecollins.graphics.components;

import static com.joecollins.graphics.utils.RenderTestUtils.compareRendering;
import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.Binding;
import com.joecollins.graphics.utils.BindableWrapper;
import java.awt.BorderLayout;
import java.awt.Color;
import java.io.IOException;
import javax.swing.JPanel;
import org.junit.Test;

public class GraphicsFrameTest {

  private static class TestObject extends Bindable {

    private enum Properties {
      NUM_POLLS
    }

    int numPolls = 0;

    public void setNumPolls(int numPolls) {
      this.numPolls = numPolls;
      onPropertyRefreshed(Properties.NUM_POLLS);
    }
  }

  @Test
  public void testFixedHeader() {
    GraphicsFrame graphicsFrame = new GraphicsFrame();
    graphicsFrame.setHeaderBinding(Binding.fixedBinding("HEADER"));
    assertEquals("HEADER", graphicsFrame.getHeader());
  }

  @Test
  public void testDynamicHeader() {
    GraphicsFrame graphicsFrame = new GraphicsFrame();
    graphicsFrame.setHeaderBinding(
        Binding.propertyBinding(
            new TestObject(),
            o -> o.numPolls + " POLLS REPORTING",
            TestObject.Properties.NUM_POLLS));
    assertEquals("0 POLLS REPORTING", graphicsFrame.getHeader());
  }

  @Test
  public void testDynamicHeaderRefreshed() {
    GraphicsFrame graphicsFrame = new GraphicsFrame();
    TestObject object = new TestObject();
    graphicsFrame.setHeaderBinding(
        Binding.propertyBinding(
            object, o -> o.numPolls + " POLLS REPORTING", TestObject.Properties.NUM_POLLS));
    assertEquals("0 POLLS REPORTING", graphicsFrame.getHeader());
    object.setNumPolls(1);
    assertEquals("1 POLLS REPORTING", graphicsFrame.getHeader());

    graphicsFrame.dispose();
    object.setNumPolls(0);
    assertEquals("1 POLLS REPORTING", graphicsFrame.getHeader());
  }

  @Test
  public void testUnbind() {
    GraphicsFrame graphicsFrame = new GraphicsFrame();
    TestObject object = new TestObject();
    graphicsFrame.setHeaderBinding(
        Binding.propertyBinding(
            object, o -> o.numPolls + " POLLS REPORTING", TestObject.Properties.NUM_POLLS));
    assertEquals("0 POLLS REPORTING", graphicsFrame.getHeader());
    graphicsFrame.setHeaderBinding(Binding.fixedBinding("FIXED BINDING"));
    assertEquals("FIXED BINDING", graphicsFrame.getHeader());
    object.setNumPolls(1);
    assertEquals("FIXED BINDING", graphicsFrame.getHeader());
  }

  @Test
  public void testFixedNotes() {
    GraphicsFrame graphicsFrame = new GraphicsFrame();
    graphicsFrame.setNotesBinding(Binding.fixedBinding("SOURCE: BBC"));
    assertEquals("SOURCE: BBC", graphicsFrame.getNotes());
  }

  @Test
  public void testBorderColor() {
    GraphicsFrame graphicsFrame = new GraphicsFrame();
    graphicsFrame.setBorderColorBinding(Binding.fixedBinding(Color.BLUE));
    assertEquals(Color.BLUE, graphicsFrame.getBorderColor());
  }

  @Test
  public void testRenderingHeaderOnly() throws IOException {
    GraphicsFrame graphicsFrame =
        new GraphicsFrame() {
          {
            JPanel panel = new JPanel();
            panel.setBackground(Color.YELLOW);
            add(panel, BorderLayout.CENTER);
          }
        };
    graphicsFrame.setHeaderBinding(Binding.fixedBinding("HEADER"));
    graphicsFrame.setSize(256, 128);
    compareRendering("GraphicsFrame", "HeaderOnly", graphicsFrame);
  }

  @Test
  public void testRenderingHeaderAndNotes() throws IOException {
    GraphicsFrame graphicsFrame =
        new GraphicsFrame() {
          {
            JPanel panel = new JPanel();
            panel.setBackground(Color.YELLOW);
            add(panel, BorderLayout.CENTER);
          }
        };
    graphicsFrame.setHeaderBinding(Binding.fixedBinding("HEADER"));
    graphicsFrame.setNotesBinding(Binding.fixedBinding("SOURCE: BBC"));
    graphicsFrame.setSize(256, 128);
    compareRendering("GraphicsFrame", "HeaderAndNotes", graphicsFrame);
  }

  @Test
  public void testRenderingNoHeader() throws IOException {
    GraphicsFrame graphicsFrame =
        new GraphicsFrame() {
          {
            JPanel panel = new JPanel();
            panel.setBackground(Color.YELLOW);
            add(panel, BorderLayout.CENTER);
          }
        };
    graphicsFrame.setSize(256, 128);
    compareRendering("GraphicsFrame", "NoHeader", graphicsFrame);
  }

  @Test
  public void testRenderingBorderColor() throws IOException {
    GraphicsFrame graphicsFrame =
        new GraphicsFrame() {
          {
            JPanel panel = new JPanel();
            panel.setBackground(Color.YELLOW);
            add(panel, BorderLayout.CENTER);
          }
        };
    graphicsFrame.setHeaderBinding(Binding.fixedBinding("HEADER"));
    graphicsFrame.setNotesBinding(Binding.fixedBinding("SOURCE: BBC"));
    graphicsFrame.setBorderColorBinding(Binding.fixedBinding(Color.RED));
    graphicsFrame.setSize(256, 128);
    compareRendering("GraphicsFrame", "BorderColor", graphicsFrame);
  }

  @Test
  public void testRenderingAccents() throws IOException {
    GraphicsFrame graphicsFrame =
        new GraphicsFrame() {
          {
            JPanel panel = new JPanel();
            panel.setBackground(Color.YELLOW);
            add(panel, BorderLayout.CENTER);
          }
        };
    graphicsFrame.setHeaderBinding(Binding.fixedBinding("\u00c7A C'EST GR\u00c2VE"));
    graphicsFrame.setNotesBinding(Binding.fixedBinding("JOYEUX NO\u00cbL, GAR\u00c7ON!"));
    graphicsFrame.setSize(256, 128);
    compareRendering("GraphicsFrame", "Accents", graphicsFrame);
  }

  @Test
  public void testHeaderFontSize() throws IOException {
    GraphicsFrame graphicsFrame =
        new GraphicsFrame() {
          {
            JPanel panel = new JPanel();
            panel.setBackground(Color.YELLOW);
            add(panel, BorderLayout.CENTER);
          }
        };
    BindableWrapper<String> headerWrapper = new BindableWrapper<>("THIS IS A HEADER");
    graphicsFrame.setHeaderBinding(headerWrapper.getBinding());
    graphicsFrame.setSize(256, 128);
    compareRendering("GraphicsFrame", "HeaderFontSize-1", graphicsFrame);

    headerWrapper.setValue("THIS IS A VERY MUCH LONGER HEADER");
    compareRendering("GraphicsFrame", "HeaderFontSize-2", graphicsFrame);

    graphicsFrame.setSize(512, 128);
    compareRendering("GraphicsFrame", "HeaderFontSize-3", graphicsFrame);
  }

  @Test
  public void testRenderHeaderAlignment() throws IOException {
    BindableWrapper<GraphicsFrame.Alignment> alignment =
        new BindableWrapper<>(GraphicsFrame.Alignment.CENTER);
    GraphicsFrame graphicsFrame =
        new GraphicsFrame() {
          {
            JPanel panel = new JPanel();
            panel.setBackground(Color.YELLOW);
            add(panel, BorderLayout.CENTER);
          }
        };
    graphicsFrame.setHeaderBinding(Binding.fixedBinding("HEADER"));
    graphicsFrame.setHeaderAlignmentBinding(alignment.getBinding());
    graphicsFrame.setSize(256, 128);
    compareRendering("GraphicsFrame", "HeaderAlignment-1", graphicsFrame);

    alignment.setValue(GraphicsFrame.Alignment.LEFT);
    compareRendering("GraphicsFrame", "HeaderAlignment-2", graphicsFrame);

    alignment.setValue(GraphicsFrame.Alignment.RIGHT);
    compareRendering("GraphicsFrame", "HeaderAlignment-3", graphicsFrame);
  }
}

File: src/test/java/com/joecollins/graphics/components/GraphicsFrameTest.kt
package com.joecollins.graphics.components

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding.Companion.fixedBinding
import com.joecollins.bindings.Binding.Companion.propertyBinding
import com.joecollins.graphics.utils.BindableWrapper
import com.joecollins.graphics.utils.RenderTestUtils.compareRendering
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.Test
import java.awt.BorderLayout
import java.awt.Color
import java.io.IOException
import java.util.concurrent.TimeUnit
import javax.swing.JPanel
import kotlin.Throws

class GraphicsFrameTest {
    private class TestObject : Bindable<TestObject, TestObject.Properties>() {
        enum class Properties {
            NUM_POLLS
        }

        private var _numPolls = 0

        var numPolls: Int
            get() = _numPolls
            set(numPolls) {
                this._numPolls = numPolls
                onPropertyRefreshed(Properties.NUM_POLLS)
            }
    }

    @Test
    fun testFixedHeader() {
        val graphicsFrame = GraphicsFrame(
            headerPublisher = fixedBinding("HEADER").toPublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ graphicsFrame.header }, IsEqual("HEADER"))
    }

    @Test
    fun testDynamicHeader() {
        val graphicsFrame = GraphicsFrame(
            headerPublisher = propertyBinding(TestObject(), { it.numPolls.toString() + " POLLS REPORTING" }, TestObject.Properties.NUM_POLLS).toPublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ graphicsFrame.header }, IsEqual("0 POLLS REPORTING"))
    }

    @Test
    fun testDynamicHeaderRefreshed() {
        val obj = TestObject()
        val graphicsFrame = GraphicsFrame(
            headerPublisher = propertyBinding(obj, { it.numPolls.toString() + " POLLS REPORTING" }, TestObject.Properties.NUM_POLLS).toPublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ graphicsFrame.header }, IsEqual("0 POLLS REPORTING"))
        obj.numPolls = 1
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ graphicsFrame.header }, IsEqual("1 POLLS REPORTING"))
    }

    @Test
    fun testFixedNotes() {
        val graphicsFrame = GraphicsFrame(
            headerPublisher = fixedBinding(null).toPublisher(),
            notesPublisher = fixedBinding("SOURCE: BBC").toPublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ graphicsFrame.notes }, IsEqual("SOURCE: BBC"))
    }

    @Test
    fun testBorderColor() {
        val graphicsFrame = GraphicsFrame(
            headerPublisher = fixedBinding(null).toPublisher(),
            borderColorPublisher = fixedBinding(Color.BLUE).toPublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ graphicsFrame.borderColor }, IsEqual(Color.BLUE))
    }

    @Test
    @Throws(IOException::class)
    fun testRenderingHeaderOnly() {
        val graphicsFrame: GraphicsFrame = object : GraphicsFrame(
            headerPublisher = fixedBinding("HEADER").toPublisher()
        ) {
            init {
                val panel = JPanel()
                panel.background = Color.YELLOW
                add(panel, BorderLayout.CENTER)
            }
        }
        graphicsFrame.setSize(256, 128)
        compareRendering("GraphicsFrame", "HeaderOnly", graphicsFrame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderingHeaderAndNotes() {
        val graphicsFrame: GraphicsFrame = object : GraphicsFrame(
            headerPublisher = fixedBinding("HEADER").toPublisher(),
            notesPublisher = fixedBinding("SOURCE: BBC").toPublisher()
        ) {
            init {
                val panel = JPanel()
                panel.background = Color.YELLOW
                add(panel, BorderLayout.CENTER)
            }
        }
        graphicsFrame.setSize(256, 128)
        compareRendering("GraphicsFrame", "HeaderAndNotes", graphicsFrame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderingNoHeader() {
        val graphicsFrame: GraphicsFrame = object : GraphicsFrame(
            headerPublisher = fixedBinding(null).toPublisher()
        ) {
            init {
                val panel = JPanel()
                panel.background = Color.YELLOW
                add(panel, BorderLayout.CENTER)
            }
        }
        graphicsFrame.setSize(256, 128)
        compareRendering("GraphicsFrame", "NoHeader", graphicsFrame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderingBorderColor() {
        val graphicsFrame: GraphicsFrame = object : GraphicsFrame(
            headerPublisher = fixedBinding("HEADER").toPublisher(),
            notesPublisher = fixedBinding("SOURCE: BBC").toPublisher(),
            borderColorPublisher = fixedBinding(Color.RED).toPublisher()
        ) {
            init {
                val panel = JPanel()
                panel.background = Color.YELLOW
                add(panel, BorderLayout.CENTER)
            }
        }
        graphicsFrame.setSize(256, 128)
        compareRendering("GraphicsFrame", "BorderColor", graphicsFrame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderingAccents() {
        val graphicsFrame: GraphicsFrame = object : GraphicsFrame(
            headerPublisher = fixedBinding("\u00c7A C'EST GR\u00c2VE").toPublisher(),
            notesPublisher = fixedBinding("JOYEUX NO\u00cbL, GAR\u00c7ON!").toPublisher()
        ) {
            init {
                val panel = JPanel()
                panel.background = Color.YELLOW
                add(panel, BorderLayout.CENTER)
            }
        }
        graphicsFrame.setSize(256, 128)
        compareRendering("GraphicsFrame", "Accents", graphicsFrame)
    }

    @Test
    @Throws(IOException::class)
    fun testHeaderFontSize() {
        val headerWrapper = BindableWrapper("THIS IS A HEADER")
        val graphicsFrame: GraphicsFrame = object : GraphicsFrame(
            headerPublisher = headerWrapper.binding.toPublisher()
        ) {
            init {
                val panel = JPanel()
                panel.background = Color.YELLOW
                add(panel, BorderLayout.CENTER)
            }
        }
        graphicsFrame.setSize(256, 128)
        compareRendering("GraphicsFrame", "HeaderFontSize-1", graphicsFrame)
        headerWrapper.value = "THIS IS A VERY MUCH LONGER HEADER"
        compareRendering("GraphicsFrame", "HeaderFontSize-2", graphicsFrame)
        graphicsFrame.setSize(512, 128)
        compareRendering("GraphicsFrame", "HeaderFontSize-3", graphicsFrame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderHeaderAlignment() {
        val alignment = BindableWrapper(GraphicsFrame.Alignment.CENTER)
        val graphicsFrame: GraphicsFrame = object : GraphicsFrame(
            headerPublisher = fixedBinding("HEADER").toPublisher(),
            headerAlignmentPublisher = alignment.binding.toPublisher()
        ) {
            init {
                val panel = JPanel()
                panel.background = Color.YELLOW
                add(panel, BorderLayout.CENTER)
            }
        }
        graphicsFrame.setSize(256, 128)
        compareRendering("GraphicsFrame", "HeaderAlignment-1", graphicsFrame)
        alignment.value = GraphicsFrame.Alignment.LEFT
        compareRendering("GraphicsFrame", "HeaderAlignment-2", graphicsFrame)
        alignment.value = GraphicsFrame.Alignment.RIGHT
        compareRendering("GraphicsFrame", "HeaderAlignment-3", graphicsFrame)
    }
}
